homeassistant/components/systemmonitor/coordinator.py

```diff
@@ -0,0 +1,233 @@
+
+
+class MonitorCoordinator(DataUpdateCoordinator[_dataT]):
+    """A System monitor Base Data Update Coordinator."""
+
+    def __init__(self, hass: HomeAssistant, type_: str, argument: str) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=f"{DOMAIN} {type_}_{argument}",
+            update_interval=DEFAULT_SCAN_INTERVAL,
+            always_update=False,
+        )
+        self._type = type_
+        self._argument = argument
+
+    async def _async_update_data(self) -> _dataT:
+        """Fetch data."""
+        return await self.hass.async_add_executor_job(self.update_data)
+
+    @abstractmethod
+    def update_data(self) -> _dataT:
+        """To be extended by data update coordinators."""
+
+
+class SystemMonitorDiskCoordinator(MonitorCoordinator[sdiskusage]):
+    """A System monitor Disk Data Update Coordinator."""
+
+    def update_data(self) -> sdiskusage:
+        """Fetch data."""
+        return psutil.disk_usage(self._argument)
+
+
+class SystemMonitorSwapCoordinator(MonitorCoordinator[sswap]):
+    """A System monitor Swap Data Update Coordinator."""
+
+    def update_data(self) -> sswap:
+        """Fetch data."""
+        return psutil.swap_memory()
+
+
+class SystemMonitorMemoryCoordinator(MonitorCoordinator[svmem]):
+    """A System monitor Memory Data Update Coordinator."""
+
+    def update_data(self) -> svmem:
+        """Fetch data."""
+        return psutil.virtual_memory()
+
+
+class SystemMonitorNetIOCoordinator(MonitorCoordinator[int]):
+    """A System monitor Network IO Data Update Coordinator."""
+
+    def update_data(self) -> int:
+        """Fetch data."""
+        counters = psutil.net_io_counters(pernic=True)
+        if self._argument in counters:
+            return counters[self._argument][IO_COUNTER[self._type]]
+        raise UpdateFailed(f"NIC {self._argument} could not be found")
+
+
+class SystemMonitorNetThroughputCoordinator(MonitorCoordinator[float | None]):
+    """A System monitor Network Throughput Data Update Coordinator."""
+
+    def __init__(self, hass: HomeAssistant, type_: str, argument: str) -> None:
+        """Initialize the coordinator."""
+        super().__init__(hass, type_, argument)
+        self._value: float | None = None
+        self._update_time: datetime | None = None
+
+    def update_data(self) -> float | None:
+        """Fetch data."""
+        counters = psutil.net_io_counters(pernic=True)
+        if self._argument in counters:
+            counter = counters[self._argument][IO_COUNTER[self._type]]
+            now = dt_util.utcnow()
+            if self._value and self._value < counter:
+                return round(
+                    (counter - self._value)
+                    / 1000**2
+                    / (now - (self._update_time or now)).total_seconds(),
+                    3,
+                )
+            self._update_time = now
+            self._value = counter
```

Comment:
Look at the renault sensor platform that passes an entity to the description value method. I think we can do it like that to be able to set entity attributes and store data that we need.

Side note: Calculating state like this isn't really allowed but we should keep it for now.